Added unit-tests for division

jeremyirilli opened the pull request:
I've added unit tests to 'interval' for the division operation.

<br>

![image](https://github.com/mgondan/interval/assets/149394139/8b8e7dbb-efca-4f06-9b2d-9ac48c4d02a7)

<br>

I used once/1 in every test to avoid this warning:

<br>

![image](https://github.com/mgondan/interval/assets/149394139/20f5216c-0f60-4c18-939a-9931a77b74b0)

<br>

Another option would have been to add a cut in the respective predicates...

I also tried to test the predicates with two solutions, such as this one:

<br>

![image](https://github.com/mgondan/interval/assets/149394139/d634b187-e084-4b68-8d8c-eaf454b6fe43)

<br>


with setof/3, but it did not work. The issue seems to be a more general one related to the fact that right now Prolog does not backtrack to find multiple solutions, even with minimal examples. 

For example, I have this module:

:- module(debug, [f/1]).

f(1).
f(2).


and I get only one solution:
?- f(X).
X = 1 .

?- 

It is very strange...

<br>

<br>

I further added two test for 'mcint' (frac and dfrac) and tried to fix the other two (omit_left and omit_right), which did not pass. 
The result of 
?- interval(omit_left(11...12 - 20...21), Res).
is
Res = -8.

and not an interval:
L > 19.9999,
L < 20.0001,
U > 20.9999,
U < 21.0001.

as defined in the  test.
But the result in test is the expected one, right?

mgondan commented:
We should get rid of these choice points, without once/1

mgondan commented:
Just to make sure, you pressed semicolon after the first response?


-------- Ursprüngliche Nachricht --------
Von: jeremyirilli ***@***.***>
Datum: 05.07.24 16:01 (GMT+01:00)
An: mgondan/interval ***@***.***>
Cc: Subscribed ***@***.***>
Betreff: [mgondan/interval] Added unit-tests (PR #3)


I've added unit test to 'interval' for the division operation.


image.png (view on web)<https://github.com/mgondan/interval/assets/149394139/8b8e7dbb-efca-4f06-9b2d-9ac48c4d02a7>


I used once/1 in every test to avoid this warning:


image.png (view on web)<https://github.com/mgondan/interval/assets/149394139/20f5216c-0f60-4c18-939a-9931a77b74b0>


Another option would have been to add a cut in the respective predicates...

I also tried to test the predicates with two solutions, such as this one:


image.png (view on web)<https://github.com/mgondan/interval/assets/149394139/d634b187-e084-4b68-8d8c-eaf454b6fe43>


with setof/3, but it did not work. The issue seems to be a more general one related to the fact that right now Prolog does not backtrack to find multiple solutions, even with minimal examples.

For example, I have this module:

:- module(debug, [f/1]).

f(1).
f(2).

and I get only one solution:
?- f(X).
X = 1 .

?-

It is very strange...



I further added two test for 'mcint' (frac and dfrac) and tried to fix the other two (omit_left and omit_right), which did not pass.
The result of
?- interval(omit_left(11...12 - 20...21), Res).
is
Res = -8.

and not an interval:
L > 19.9999,
L < 20.0001,
U > 20.9999,
U < 21.0001.

as defined in the test.
But the result in test is the expected one, right?

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/mgondan/interval/pull/3

Commit Summary

  *   9b766d3<https://github.com/mgondan/interval/pull/3/commits/9b766d3da83060b95050b6034ce9d18bbe56c362> Added unit-tests for division
  *   2bb5c01<https://github.com/mgondan/interval/pull/3/commits/2bb5c0168b8729714d7eb24afd87e51523336c61> Added unit tests mcclass fractions

File Changes

(3 files<https://github.com/mgondan/interval/pull/3/files>)

  *   M prolog/mcclass.pl<https://github.com/mgondan/interval/pull/3/files#diff-2a37517dd680367883770db212484e8e1638f2f88b65a015fd5fa7da97e9ef7b> (2)
  *   M test/test_interval.pl<https://github.com/mgondan/interval/pull/3/files#diff-898a14a6a8d12fe351a821f719b863a65353ae03778fe988c3e1b1a42bb8664c> (201)
  *   M test/test_mcclass.pl<https://github.com/mgondan/interval/pull/3/files#diff-8507b9188b698c6ad3926fcd39a4d07f7ee55fde73045e85646c5af128893f1e> (16)

Patch Links:

  *   https://github.com/mgondan/interval/pull/3.patch
  *   https://github.com/mgondan/interval/pull/3.diff

—
Reply to this email directly, view it on GitHub<https://github.com/mgondan/interval/pull/3>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADABF6UJYPCUUKSKOATFBZLZK2RMZAVCNFSM6AAAAABKNIKPFKVHI2DSMVQWIX3LMV43ASLTON2WKOZSGM4TENRXGEZDOMA>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

mgondan commented:
interval(omit_left(11...12 - 20...21), Res)

Should return 20...21

jeremyirilli commented:
> We should get rid of these choice points, without once/1

Ok, then I will add a cut in the respective predicates.

mgondan commented:
Please do it in separate PRs for each operation (e.g., one PR for division).

Von: jeremyirilli ***@***.***>
Gesendet: Freitag, 5. Juli 2024 20:40
An: mgondan/interval ***@***.***>
Cc: Gondan-Rochon, Matthias ***@***.***>; Comment ***@***.***>
Betreff: Re: [mgondan/interval] Added unit-tests (PR #3)


We should get rid of these choice points, without once/1

Ok, then I will add a cut in the respective predicates.

—
Reply to this email directly, view it on GitHub<https://github.com/mgondan/interval/pull/3#issuecomment-2211275147>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADABF6UCJ4MXVNWCMVNB2HDZK3R6JAVCNFSM6AAAAABKNIKPFKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMJRGI3TKMJUG4>.
You are receiving this because you commented.Message ID: ***@***.******@***.***>>

jeremyirilli commented:
> Just to make sure, you pressed semicolon after the first response?
> […](#)
> -------- Ursprüngliche Nachricht -------- Von: jeremyirilli ***@***.***> Datum: 05.07.24 16:01 (GMT+01:00) An: mgondan/interval ***@***.***> Cc: Subscribed ***@***.***> Betreff: [mgondan/interval] Added unit-tests (PR #3) I've added unit test to 'interval' for the division operation. image.png (view on web)<https://github.com/mgondan/interval/assets/149394139/8b8e7dbb-efca-4f06-9b2d-9ac48c4d02a7> I used once/1 in every test to avoid this warning: image.png (view on web)<https://github.com/mgondan/interval/assets/149394139/20f5216c-0f60-4c18-939a-9931a77b74b0> Another option would have been to add a cut in the respective predicates... I also tried to test the predicates with two solutions, such as this one: image.png (view on web)<https://github.com/mgondan/interval/assets/149394139/d634b187-e084-4b68-8d8c-eaf454b6fe43> with setof/3, but it did not work. The issue seems to be a more general one related to the fact that right now Prolog does not backtrack to find multiple solutions, even with minimal examples. For example, I have this module: :- module(debug, [f/1]). f(1). f(2). and I get only one solution: ?- f(X). X = 1 . ?- It is very strange... I further added two test for 'mcint' (frac and dfrac) and tried to fix the other two (omit_left and omit_right), which did not pass. The result of ?- interval(omit_left(11...12 - 20...21), Res). is Res = -8. and not an interval: L > 19.9999, L < 20.0001, U > 20.9999, U < 21.0001. as defined in the test. But the result in test is the expected one, right?
> ________________________________ You can view, comment on, or merge this pull request online at: #3 Commit Summary * [9b766d3](https://github.com/mgondan/interval/commit/9b766d3da83060b95050b6034ce9d18bbe56c362)<[9b766d3](https://github.com/mgondan/interval/pull/3/commits/9b766d3da83060b95050b6034ce9d18bbe56c362)> Added unit-tests for division * [2bb5c01](https://github.com/mgondan/interval/commit/2bb5c0168b8729714d7eb24afd87e51523336c61)<[2bb5c01](https://github.com/mgondan/interval/pull/3/commits/2bb5c0168b8729714d7eb24afd87e51523336c61)> Added unit tests mcclass fractions File Changes (3 files<https://github.com/mgondan/interval/pull/3/files>) * M prolog/mcclass.pl<https://github.com/mgondan/interval/pull/3/files#diff-2a37517dd680367883770db212484e8e1638f2f88b65a015fd5fa7da97e9ef7b> (2) * M test/test_interval.pl<https://github.com/mgondan/interval/pull/3/files#diff-898a14a6a8d12fe351a821f719b863a65353ae03778fe988c3e1b1a42bb8664c> (201) * M test/test_mcclass.pl<https://github.com/mgondan/interval/pull/3/files#diff-8507b9188b698c6ad3926fcd39a4d07f7ee55fde73045e85646c5af128893f1e> (16) Patch Links: * https://github.com/mgondan/interval/pull/3.patch * https://github.com/mgondan/interval/pull/3.diff — Reply to this email directly, view it on GitHub<#3>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADABF6UJYPCUUKSKOATFBZLZK2RMZAVCNFSM6AAAAABKNIKPFKVHI2DSMVQWIX3LMV43ASLTON2WKOZSGM4TENRXGEZDOMA>. You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

No, I just pressed Enter, but now with semi-colon it works. I was thinking to much on the way it worked with Enter on SWISH-Prolog...

jeremyirilli commented:
- Removed the once/1.
- Added cuts.
- Checked for multiple solutions in the tests.

- Removed the commit about test_mcclass from this request, so that it is only about the division.

mgondan commented:
check guards, then cut, then unify result, i.e.

int_hook(A1...A2 / B1...B2, Res, _) :-
    (
        strictneg(A1, A2);
        strictpos(A1, A2)
    ),
    mixed(B1, B2),
    !,
    div(A1...A2, B1...B2, Res).

And especially here,

int_hook(A1...A2 / B1...B2, Res, _) :-
    !,
    div(A1...A2, B1...B2, Res).

If the div could fail, like this:

int_hook(A1...A2 / B1...B2, Res, _) :-
    div(A1...A2, B1...B2, R),
    !,
    Res = R.

jeremyirilli commented:
Updated the position of the cut. In that way the clause 

int_hook(A1...A2 / B1...B2, Res, _) :-
(
strictneg(A1, A2);
strictpos(A1, A2)
),
mixed(B1, B2),
!,
div(A1...A2, B1...B2, Res).

intended for multiple solutions, is not needed anymore.

jeremyirilli commented:
Changed the position of the cut and now one clause I added previously is not needed anymore.
